// ngx-filesaver/bundles/ngx-filesaver.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common/http'),exports, require('@angular/core'), require('file-saver'), require('@angular/common/http'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ngxFileSaver', ['@angular/core','@angular/common/http','exports', '@angular/core', 'file-saver', '@angular/common/http', 'rxjs'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common.http,global.ngxFileSaver = {}, global.ng.core, global.saveAs, global.ng.common.http, global.rxjs));
}(this, function (ɵngcc0,ɵngcc1,exports, core, fileSaver, http, rxjs) { 'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var FileSaverService = /** @class */ (function () {
        function FileSaverService() {
        }
        Object.defineProperty(FileSaverService.prototype, "isFileSaverSupported", {
            get: function () {
                try {
                    return !!new Blob();
                }
                catch (e) {
                    return false;
                }
            },
            enumerable: true,
            configurable: true
        });
        FileSaverService.prototype.genType = function (fileName) {
            if (!fileName || fileName.lastIndexOf('.') === -1) {
                return 'text/plain';
            }
            var type = fileName.substr(fileName.lastIndexOf('.') + 1);
            switch (type) {
                case 'txt':
                    return 'text/plain';
                case 'xml':
                case 'html':
                    return "text/" + type;
                case 'json':
                    return 'octet/stream';
                default:
                    return "application/" + type;
            }
        };
        FileSaverService.prototype.save = function (blob, fileName, filtType, option) {
            if (!blob) {
                throw new Error('Data argument should be a blob instance');
            }
            var file = new Blob([blob], { type: filtType || blob.type || this.genType(fileName) });
            fileSaver.saveAs(file, decodeURI(fileName || 'download'), option);
        };
        FileSaverService.prototype.saveText = function (txt, fileName, option) {
            var blob = new Blob([txt]);
            this.save(blob, fileName, null, option);
        };
        FileSaverService.ngInjectableDef = core.ɵɵdefineInjectable({ factory: function FileSaverService_Factory() { return new FileSaverService(); }, token: FileSaverService, providedIn: "root" });
FileSaverService.ɵfac = function FileSaverService_Factory(t) { return new (t || FileSaverService)(); };
FileSaverService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: FileSaverService, factory: function (t) { return FileSaverService.ɵfac(t); }, providedIn: 'root' });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FileSaverService, [{
        type: core.Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
        return FileSaverService;
    }());

    var __decorate$1 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var FileSaverDirective = /** @class */ (function () {
        function FileSaverDirective(el, fss, httpClient) {
            this.el = el;
            this.fss = fss;
            this.httpClient = httpClient;
            this.method = 'GET';
            this.success = new core.EventEmitter();
            this.error = new core.EventEmitter();
            if (!fss.isFileSaverSupported) {
                el.nativeElement.classList.add("filesaver__not-support");
            }
        }
        FileSaverDirective.prototype.getName = function (res) {
            return decodeURI(this.fileName || res.headers.get('filename') || res.headers.get('x-filename'));
        };
        FileSaverDirective.prototype._click = function () {
            var _this = this;
            if (!this.fss.isFileSaverSupported) {
                return;
            }
            var req = this.http;
            if (!req) {
                var params = new http.HttpParams();
                var query = this.query || {};
                // tslint:disable-next-line:forin
                for (var item in query) {
                    params.set(item, query[item]);
                }
                req = this.httpClient.request(this.method, this.url, {
                    observe: 'response',
                    responseType: 'blob',
                    headers: this.header,
                    params: params,
                });
            }
            this.setDisabled(true);
            req.subscribe(function (res) {
                if (res.status !== 200 || res.body.size <= 0) {
                    _this.error.emit(res);
                    return;
                }
                _this.fss.save(res.body, _this.getName(res), null, _this.fsOptions);
                _this.success.emit(res);
            }, function (err) { return _this.error.emit(err); }, function () { return _this.setDisabled(false); });
        };
        FileSaverDirective.prototype.setDisabled = function (status) {
            var el = this.el.nativeElement;
            el.disabled = status;
            el.classList[status ? 'add' : 'remove']("filesaver__disabled");
        };
        FileSaverDirective.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: FileSaverService },
            { type: http.HttpClient }
        ]; };
        __decorate$1([
            core.Input(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "method", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", rxjs.Observable)
        ], FileSaverDirective.prototype, "http", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "query", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "header", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", String)
        ], FileSaverDirective.prototype, "url", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", String)
        ], FileSaverDirective.prototype, "fileName", void 0);
        __decorate$1([
            core.Input(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "fsOptions", void 0);
        __decorate$1([
            core.Output(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "success", void 0);
        __decorate$1([
            core.Output(),
            __metadata("design:type", Object)
        ], FileSaverDirective.prototype, "error", void 0);
        __decorate$1([
            core.HostListener('click'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], FileSaverDirective.prototype, "_click", null);
        FileSaverDirective = __decorate$1([ __metadata("design:paramtypes", [core.ElementRef, FileSaverService, http.HttpClient])
        ], FileSaverDirective);
FileSaverDirective.ɵfac = function FileSaverDirective_Factory(t) { return new (t || FileSaverDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(FileSaverService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.HttpClient)); };
FileSaverDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: FileSaverDirective, selectors: [["", "fileSaver", ""]], hostBindings: function FileSaverDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function FileSaverDirective_click_HostBindingHandler() { return ctx._click(); });
    } }, inputs: { method: "method", http: "http", query: "query", header: "header", url: "url", fileName: "fileName", fsOptions: "fsOptions" }, outputs: { success: "success", error: "error" }, exportAs: ["fileSaver"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FileSaverDirective, [{
        type: core.Directive,
        args: [{
                selector: '[fileSaver]',
                exportAs: 'fileSaver'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: FileSaverService }, { type: ɵngcc1.HttpClient }]; }, { method: [{
            type: core.Input
        }], success: [{
            type: core.Output
        }], error: [{
            type: core.Output
        }], _click: [{
            type: core.HostListener,
            args: ['click']
        }], http: [{
            type: core.Input
        }], query: [{
            type: core.Input
        }], header: [{
            type: core.Input
        }], url: [{
            type: core.Input
        }], fileName: [{
            type: core.Input
        }], fsOptions: [{
            type: core.Input
        }] }); })();
        return FileSaverDirective;
    }());

    var __decorate$2 = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var FileSaverModule = /** @class */ (function () {
        function FileSaverModule() {
        }
FileSaverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FileSaverModule });
FileSaverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FileSaverModule_Factory(t) { return new (t || FileSaverModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FileSaverModule, { declarations: [FileSaverDirective], exports: [FileSaverDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FileSaverModule, [{
        type: core.NgModule,
        args: [{
                declarations: [FileSaverDirective],
                exports: [FileSaverDirective]
            }]
    }], function () { return []; }, null); })();
        return FileSaverModule;
    }());

    exports.FileSaverDirective = FileSaverDirective;
    exports.FileSaverModule = FileSaverModule;
    exports.FileSaverService = FileSaverService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
